osu.Game/Storyboards/Storyboard.cs

```diff
@@ -16,6 +16,7 @@ public class Storyboard
         public IEnumerable<StoryboardLayer> Layers => layers.Values;
 
         public BeatmapInfo BeatmapInfo = new BeatmapInfo();
+        public IBeatmap Beatmap { get; set; } = new Beatmap();
```

Comment:
This is kinda weird...

Reply:
So is everything else being populated with empty objects in this class. I was just following precedent, if you will.

Reply:
My problem here is two-fold. The first is that I'm struggling to follow the code to even ensure this is being written to right now, because it doesn't look like it is? The second is I think this goes above the "everything else" for me.

One thing I thought about but hesitated from posting is to split the properties out into a `BeatmapProperties` or maybe even into `BeatmapMetadata`. It feels wrong to be passing `Beatmap` around which historically has been used very close to gameplay...

To go further, both `BeatmapInfo` and `Beatmap` look wrong to me.
- If we're not going to have a general "class of info" that we're passing around, then I think `EpilepsyWarning` should be moved into `Storyboard`.
- The remaining references to `BeatmapInfo` need to be removed - it looks like there's two:
  - `ReplacesBackground` (`BeatmapInfo.Metadata.BackgroundFile`) - probably `BackgroundFile` should be duplicated as a property here.
  - `GetStoragePathFromStoryboardPath` the directionality looks inverted. It should be `DrawableStoryboard` doing the lookups on resource stores, not what is effectively a model class otherwise.

Perhaps this would be best as a pre-pass.

Reply:
> One thing I thought about but hesitated from posting is to split the properties out into a `BeatmapProperties` or maybe even into `BeatmapMetadata`

That's what https://github.com/ppy/osu/pull/23418 tried to do but [got stunlocked on naming](https://github.com/ppy/osu/pull/23418#discussion_r1187371707). If we can agree on a name I don't see why not.

Reply:
> One thing I thought about but hesitated from posting is to split the properties out into a BeatmapProperties or maybe even into BeatmapMetadata. It feels wrong to be passing Beatmap around which historically has been used very close to gameplay...

I get this, but is there an actually *requirement* for the split? In stable, for instance, we have an `.osu` file. And we have a `Beatmap`. And they are 1:1. Should we not just stick with that for simplicity? Am I missing a case where passing "too much" in one class is an issue?

Also I'm still confused. I can't find any write usages of `Storyboard.Beatmap` or its properties (except one test), even though it's read from. @bdach can you clarify here?

![JetBrains Rider-EAP 2024-07-23 at 03 22 39](https://github.com/user-attachments/assets/e5221819-319a-4505-80be-4f01a8c19be4)

![JetBrains Rider-EAP 2024-07-23 at 03 23 24](https://github.com/user-attachments/assets/45e7195f-d1a4-4c04-92b7-5d1e47215aa4)

Reply:
Judging from `master` code, it looks like the `Beatmap` would need to be set around here to work correctly:

https://github.com/ppy/osu/blob/3978d4babb86917150c7bfb5a19ce3c72830b885/osu.Game/Beatmaps/WorkingBeatmap.cs#L65

Reply:
> Also I'm still confused. I can't find any write usages of `Storyboard.Beatmap` or its properties (except one test), even though it's read from.

Yeah that's just a bug. Good spot, although I did try to preempt this PR by saying it's basically a draft of the general shape with no testing other than checking that unit tests pass.

> Judging from `master` code, it looks like the `Beatmap` would need to be set around here to work correctly:

Looks close, but the reality is uglier than that; see https://github.com/ppy/osu/pull/28473/commits/1d4d8063622dbc552fd695bfb6561fbe14b63e57. tl;dr is that storyboards go through a complete separate decoding path:

https://github.com/ppy/osu/blob/40b171f5449776b5f7a52f103c5b9ea1f16c0907/osu.Game/Beatmaps/WorkingBeatmapCache.cs#L257-L312

and thus all `IBeatmap` properties that `Storyboard` needs to read correctly _must be populated by_ `LegacyStoryboardDecoder`.

I guess this is a relevant point of discussion to the general direction here, and I'm not sure what to make of that revelation. In a way this may be more correct because this separate decoding path reads both the `.osb` (if present) and the `.osu` as a fallback, so the resultant behaviour _may_ be more correct here (needs stable cross-check) - but that's also a weak edge case that probably has never been an issue until now

The fact that decoding would need to be split and/or duplicated based on what beatmaps care about and storyboards care about could be seen as bad. Although that is arguably already a problem because that exists:

https://github.com/ppy/osu/blob/40b171f5449776b5f7a52f103c5b9ea1f16c0907/osu.Game/Beatmaps/Formats/LegacyStoryboardDecoder.cs#L72-L74